Timing Service now measures time not running a Module

Dr15Jones opened the pull request:
#### PR description:

The built in timing service now calculates the amount of time all cores are not running a module (i.e. Source, EDProducer, EDFilter, EDAnalyzer, OutputModule, ESModule or ESSource) during the data processing loop. This gets reported in the summary and to the framework job report. Time is summed across all threads.

Additional changes:
- Moved postESSyncIOV signal to run right after ESSources synchronize. This now properly encapsulate the time calling the ESSources.
- Added beginProcessing and endProcessing signals to specify when starting/ending processing of source transitions.

#### PR validation:

Framework unit tests pass.

Dr15Jones committed:
Added measurement of time spent without a task in Timing

Dr15Jones committed:
Moved postESSyncIOV signal to run right after ESSources synchronize

Previously the signal happened after all Records were asynchronously updated to their new IOVs.
Now the signals focus on the timings of calls to the ESSources.

Dr15Jones committed:
Include task time within EventSetup modules

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35015

- This PR adds an extra 88KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35015/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35015/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Dr15Jones committed:
Code format and cleanup

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- FWCore/Services (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@makortel, @missirol, @wddgit, @fwyzard this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d35193/31772/summary.html
**COMMIT**: f2a7f8f487847785a49c8fa17bd3039f133a16cf
**CMSSW**: CMSSW_13_1_X_2023-04-03-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41262/31772/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

makortel commented on FWCore/Services/plugins/Timing.cc:
```diff
@@ -469,6 +524,7 @@ namespace edm {
                                  << " - Total init:  " << total_initialization_time << "\n"
                                  << " - Total job:   " << total_job_time << "\n"
                                  << " - Total EventSetup: " << accumulatedEventSetupModuleTimings_.load() << "\n"
+                                 << " - Total non-task: " << total_time_without_tasks_ << "\n"
```
Or
```suggestion
                                 << " - Total non-module: " << total_time_without_tasks_ << "\n"
```
?

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35017

- This PR adds an extra 88KB to repository

cmsbuild commented:
Pull request #41262 was updated. @makortel, @smuzaffar, @Dr15Jones can you please check and sign again.

makortel commented on FWCore/Services/plugins/Timing.cc:
```diff
@@ -355,6 +367,48 @@ namespace edm {
+    }
+
+    void Timing::setTaskCallbacks(ActivityRegistry& iRegistry) {
+      iRegistry.preSourceSignal_.connect([this](auto) { addTask(); });
```
Why `preSourceSignal_.connect()` instead of the "more standard" `watchPreSourceEvent()`?

Why register separate callbacks for most of the signals, but not for ESModule?

Dr15Jones replied:
> Why preSourceSignal_.connect() instead of the "more standard" watchPreSourceEvent()?

The `watch` style methods require a new member function of the Service to be written while the `connect` allow the use of lambdas.

Dr15Jones replied:
> Why register separate callbacks for most of the signals, but not for ESModule?

The ESModule callbacks are always used, regardless of the options chosen while many of the other ones can be switched off.

makortel replied:
> > Why preSourceSignal_.connect() instead of the "more standard" watchPreSourceEvent()?
> 
> The `watch` style methods require a new member function of the Service to be written while the `connect` allow the use of lambdas.

I see `watch` style functions being used with lambdas, e.g. here
https://github.com/cms-sw/cmssw/blob/a9d66506bc61716e40a095faf6883fa8795b76e0/FWCore/Services/plugins/ZombieKillerService.cc#L65-L66

makortel reviewed (commented):
What about modules that launch additional tasks, like OutputModule or modules using `tbb::parallel_for()`? IIUC, the time spent in such tasks that are run (by TBB) in a different thread than the module would get counted as non-task (or non-module) time.

I still see some interest in the non-task/non-module time, but (unless I'm missing something) maybe it is not that useful to spot problems in production in the end?

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d35193/31773/summary.html
**COMMIT**: 3b718c9a6dd27038351e8342158981f68f83bff8
**CMSSW**: CMSSW_13_1_X_2023-04-03-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41262/31773/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 10 lines to the logs
 - Reco comparison results: 11 differences found in the comparisons
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3364817
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3364783
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 202 log files, 155 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-03-1100+d35193/56389/triggerResults): no differences found

Dr15Jones commented:
> What about modules that launch additional tasks, like OutputModule or modules using tbb::parallel_for()? IIUC, the time spent in such tasks that are run (by TBB) in a different thread than the module would get counted as non-task (or non-module) time.

That is true. But if we continue to run as many threads as streams, then the way we have the code built starting new modules will always be a higher priority than stealing a task from another thread (i.e. running a task from the OutputModule or tab::parallel_for). The only time task stealing will be appreciable is if the framework doesn't have enough work to schedule. Such cases I think are interesting to spot anyway.  

Also, the _time not in modules_ is not meant to indicate a problem. Instead, if and only if efficiency is low would we look at that new time to see if maybe we are having scheduling difficulties.

Dr15Jones committed:
Changes based on review comments

- Changed name of new measurement to non-module
- Use 'watch' form of signal callback functions
- Account for time in Acquire

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35055

- This PR adds an extra 16KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35055/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35055/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35056

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #41262 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

Dr15Jones commented:
Still need to write a unit test

Dr15Jones committed:
Start the loop timing at post begin ProcessBlock

This is the last transition we presently run just before going into the standard data processing loop.
This will now have the begin Stream and the begin ProcessBlock time be accounted for in the 'init'.

Dr15Jones committed:
Add unit test for Timing service output

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35076

- This PR adds an extra 92KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Services/test/BuildFile.xml  modified in PR(s): #41248

cmsbuild commented:
Pull request #41262 was updated. @makortel, @smuzaffar, @Dr15Jones can you please check and sign again.

makortel commented on FWCore/Services/plugins/Timing.cc:
```diff
@@ -220,8 +230,8 @@ namespace edm {
       iRegistry.watchPreBeginJob(this, &Timing::preBeginJob);
-      iRegistry.watchPostBeginJob(this, &Timing::postBeginJob);
-      iRegistry.preEndJobSignal_.connect([this]() {
+      iRegistry.watchPostBeginProcessBlock(this, &Timing::postBeginProcessBlock);
```
@wddgit Could you confirm that the `postBeginProcessBlockSignal_` is issued only for the beginning of the ProcessBlock of the current process, and not for the ProcessBlocks stored in the input file? (I believe this is the case, but want to be sure)

makortel replied:
I'm more after that the ProcessBlocks of the earlier processes (from the input file) do not cause this signal to be issued.

wddgit replied:
Yes, I confirm that is correct. There is a separate signal for the input process block transitions.

wddgit replied:
I have not read the whole PR carefully and I am not sure if this is relevant to this PR, but possibly this is worth mentioning. There are some corner cases where there can be more than one begin ProcessBlock transition in the same process. From the TWIKI:

"Normally, there is one beginProcessBlock and one endProcessBlock transition in a process. There can be more than one of each in the exceptional cases of EDLooper, SubProcess, and single OutputModules that are configured to write multiple files."

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d35193/31850/summary.html
**COMMIT**: 06390764b834ab49120227a3f672e2f0b8d880da
**CMSSW**: CMSSW_13_1_X_2023-04-05-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41262/31850/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 26 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3365163
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3365135
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 202 log files, 155 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-05-1100+d35193/56435/triggerResults): no differences found

Dr15Jones committed:
Add beginProcessing and endProcessing signals

Add explicit signals to exactly specify when initialization has finished and when shutdown begins.

Dr15Jones committed:
Use new beginProcessing signal with Timing

Dr15Jones commented:
To avoid confusion with the exact timing of beginning to process source transitions, I added two new signals into the system: `beginProcessing` and `endProcessing`. These specifically denote when we are beginning to start the processing loop and when we have finished the processing loop.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35079

- This PR adds an extra 104KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Services/test/BuildFile.xml  modified in PR(s): #41248

cmsbuild commented:
Pull request #41262 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

makortel commented on FWCore/ServiceRegistry/interface/ActivityRegistry.h:
```diff
@@ -327,6 +327,18 @@ namespace edm {
+    AR_WATCH_USING_METHOD_0(watchBeginProcessing)
+
+    typedef signalslot::Signal<void()> EndProcessing;
+    /// signal is emitted after all work has been done processing all transitions
```
I suppose "all transitions" includes endProcessBlock, but not endJob(+endStream). How about clarifying along "all transitions except endJob and endStream"?

(otherwise I think the separate signal is a good clarification)

Dr15Jones replied:
changed to `processing all source transitions`

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41262/35104

- This PR adds an extra 20KB to repository

Dr15Jones commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d35193/31879/summary.html
**COMMIT**: 16dd97839bcb303fe8f637237c15954ada19aa73
**CMSSW**: CMSSW_13_1_X_2023-04-06-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41262/31879/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 5 lines to the logs
 - Reco comparison results: 14 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3459609
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3459578
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-06-1100+d35193/56456/triggerResults): no differences found

makortel commented:
@Dr15Jones Could you add the other changes of the PR (the moment where "init" accounting changes to "loop" accounting, change in where `postESSyncIOVSignal_` is issued) to the PR description?

makortel commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

slava77 commented:
will this PR allow counting the time spent on GPU as well? IIUC, currently it's not included in the TimingService per-module totals.

makortel commented:
> will this PR allow counting the time spent on GPU as well? IIUC, currently it's not included in the TimingService per-module totals.

No. Already the exact meaning of "time spent on GPU" is complex enough to warrant its own issue and discussion (although https://github.com/cms-sw/cmssw/issues/29142 is somewhat related).

rappoccio commented:
+1

slava77 commented:
> > will this PR allow counting the time spent on GPU as well? IIUC, currently it's not included in the TimingService per-module totals.
> 
> No. Already the exact meaning of "time spent on GPU" is complex enough to warrant its own issue and discussion (although #29142 is somewhat related).

what about at least in a single thread/stream execution case?

makortel commented:
> > > will this PR allow counting the time spent on GPU as well? IIUC, currently it's not included in the TimingService per-module totals.
> > 
> > 
> > No. Already the exact meaning of "time spent on GPU" is complex enough to warrant its own issue and discussion (although #29142 is somewhat related).
> 
> what about at least in a single thread/stream execution case?

I can imagine several possible things to measure as "time spent on GPU" even then, so I'd rather figure out first what quantity/quantities exactly make sense to measure (suggestions are welcome in a separate issye), and then craft a general solution for that.

But I acknowledge that a long running "external worker" is indeed one source of "non-module time" in a situation where the framework('s thread) has nothing else to do than wait for the work to complete.
